--- a/arch/mips/ath79/mach-ap136.c
+++ b/arch/mips/ath79/mach-ap136.c
@@ -131,7 +131,8 @@
 static struct mdio_board_info ap136_mdio0_info[] = {
 	{
 		.bus_id = "ag71xx-mdio.0",
-		.phy_addr = 0,
+		.modalias = "ag71xx-mdio",
+		.mdio_addr = 0,
 		.platform_data = &ap136_ar8327_data,
 	},
 };
--- a/arch/mips/ath79/mach-db120.c
+++ b/arch/mips/ath79/mach-db120.c
@@ -124,7 +124,8 @@
 static struct mdio_board_info db120_mdio0_info[] = {
 	{
 		.bus_id = "ag71xx-mdio.0",
-		.phy_addr = 0,
+		.modalias = "ag71xx-mdio",
+		.mdio_addr = 0,
 		.platform_data = &db120_ar8327_data,
 	},
 };
