diff --git a/ecm_conntrack_notifier.c b/ecm_conntrack_notifier.c
index 59a291e..f691f07 100644
--- a/ecm_conntrack_notifier.c
+++ b/ecm_conntrack_notifier.c
@@ -317,15 +317,9 @@ EXPORT_SYMBOL(ecm_conntrack_ipv4_event);
  * ecm_conntrack_event()
  *	Callback event invoked when conntrack connection state changes, currently we handle destroy events to quickly release state
  */
-#ifdef CONFIG_NF_CONNTRACK_CHAIN_EVENTS
 static int ecm_conntrack_event(struct notifier_block *this, unsigned long events, void *ptr)
-#else
-static int ecm_conntrack_event(unsigned int events, struct nf_ct_event *item)
-#endif
 {
-#ifdef CONFIG_NF_CONNTRACK_CHAIN_EVENTS
 	struct nf_ct_event *item = (struct nf_ct_event *)ptr;
-#endif
 	struct nf_conn *ct = item->ct;
 
 	/*
@@ -358,7 +352,6 @@ static int ecm_conntrack_event(unsigned int events, struct nf_ct_event *item)
 	return NOTIFY_DONE;
 }
 
-#ifdef CONFIG_NF_CONNTRACK_CHAIN_EVENTS
 /*
  * struct notifier_block ecm_conntrack_notifier
  *	Netfilter conntrack event system to monitor connection tracking changes
@@ -366,15 +359,6 @@ static int ecm_conntrack_event(unsigned int events, struct nf_ct_event *item)
 static struct notifier_block ecm_conntrack_notifier = {
 	.notifier_call	= ecm_conntrack_event,
 };
-#else
-/*
- * struct nf_ct_event_notifier ecm_conntrack_notifier
- *	Netfilter conntrack event system to monitor connection tracking changes
- */
-static struct nf_ct_event_notifier ecm_conntrack_notifier = {
-	.fcn	= ecm_conntrack_event,
-};
-#endif
 #endif
 
 /*
